### ISSUE TITLE
You requested storage driver docker_registry.drivers.swift which is not installed

### ISSUE THREAD

**Issue by agonzalezandino** [bacongobbler/docker-registry-driver-swift#23]
Hello!

I'm looking forward to using your driver in order to privately store images in Rackspace Cloudfiles. But I've found a problem during the process: docker is installed and I've successfully compiled the driver by running pip install docker-registry-driver-swift.

```
# pip freeze
Babel==1.3
boto==2.34.0
chardet==2.0.1
Cheetah==2.4.4
cloud-init==0.7.2
configobj==4.7.2
distro-info==0.10
docker-registry-core==2.0.3
docker-registry-driver-swift==0.2.1
euca2ools==2.0.2
gevent==0.13.6
GnuPGInterface==0.3.2
greenlet==0.3.1
iso8601==0.1.10
lxml==3.4.1
M2Crypto==0.21.1
medusa==0.5.4
meld3==0.6.5
netaddr==0.7.13
oauth==1.0.1
oauthlib==0.1.2
oslo.config==1.6.0
paramiko==1.7.7.1
pbr==0.10.7
prettytable==0.7.2
pycrypto==2.6
pycurl==7.19.0
python-apt==0.8.8.2
python-keystoneclient==0.10.0
python-swiftclient==2.2.0
pytz==2014.10
PyYAML==3.10
redis==2.10.3
requests==2.5.1
simplejson==3.6.2
six==1.9.0
stevedore==1.2.0
supervisor==3.0a8
ufw==0.31.1.post2
unattended-upgrades==0.1
```

Ok, everything seems to be ready to launch the container with:

```
docker run \
  -e LOGLEVEL=INFO \
  -e SETTINGS_FLAVOR=swift \
  -e STORAGE_PATH=/mypath \
  -e OS_CONTAINER=mycontainer \
  -e OS_USERNAME=myuser \
  -e OS_PASSWORD=mypassword \
  -e OS_REGION_NAME=LON \
  -e OS_AUTH_URL="https://identity.api.rackspacecloud.com/v2.0" \
  -e SEARCH_BACKEND="sqlalchemy" \
  -p 5000:5000 \
  registry
```

And this is the awful output that I'm getting:

```
[2015-01-15 20:39:33 +0000] [1] [INFO] Starting gunicorn 19.1.1
[2015-01-15 20:39:33 +0000] [1] [INFO] Listening at: http://0.0.0.0:5000 (1)
[2015-01-15 20:39:33 +0000] [1] [INFO] Using worker: gevent
[2015-01-15 20:39:33 +0000] [13] [INFO] Booting worker with pid: 13
[2015-01-15 20:39:34 +0000] [16] [INFO] Booting worker with pid: 16
[2015-01-15 20:39:34 +0000] [17] [INFO] Booting worker with pid: 17
15/Jan/2015:20:39:34 +0000 WARNING: Cache storage disabled!
15/Jan/2015:20:39:34 +0000 WARNING: LRU cache disabled!
15/Jan/2015:20:39:34 +0000 WARNING: Got exception: No module named swift
[2015-01-15 20:39:34 +0000] [13] [ERROR] Exception in worker process:
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/arbiter.py", line 507, in spawn_worker
    worker.init_process()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/workers/ggevent.py", line 193, in init_process
    super(GeventWorker, self).init_process()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/workers/base.py", line 114, in init_process
    self.wsgi = self.app.wsgi()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/base.py", line 66, in wsgi
    self.callable = self.load()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/wsgiapp.py", line 65, in load
    return self.load_wsgiapp()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/wsgiapp.py", line 52, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/util.py", line 356, in import_app
    __import__(module)
  File "/usr/local/lib/python2.7/dist-packages/docker_registry/wsgi.py", line 20, in <module>
    from .tags import *  # noqa
  File "/usr/local/lib/python2.7/dist-packages/docker_registry/tags.py", line 22, in <module>
    store = storage.load()
  File "/usr/local/lib/python2.7/dist-packages/docker_registry/storage/__init__.py", line 36, in load
    _storage[kind] = engine.fetch(kind)(
  File "/usr/local/lib/python2.7/dist-packages/docker_registry/core/driver.py", line 267, in fetch
    % (name, name, available(), e)
NotImplementedError: You requested storage driver docker_registry.drivers.swift
which is not installed. Try `pip install docker-registry-driver-swift`
or check your configuration. The following are currently
available on your system: ['dumb', 'file', 's3']. Exception was: No module named swift
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/arbiter.py", line 507, in spawn_worker
    worker.init_process()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/workers/ggevent.py", line 193, in init_process
    super(GeventWorker, self).init_process()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/workers/base.py", line 114, in init_process
    self.wsgi = self.app.wsgi()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/base.py", line 66, in wsgi
    self.callable = self.load()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/wsgiapp.py", line 65, in load
    return self.load_wsgiapp()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/wsgiapp.py", line 52, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/util.py", line 356, in import_app
    __import__(module)
  File "/usr/local/lib/python2.7/dist-packages/docker_registry/wsgi.py", line 20, in <module>
    from .tags import *  # noqa
  File "/usr/local/lib/python2.7/dist-packages/docker_registry/tags.py", line 22, in <module>
    store = storage.load()
  File "/usr/local/lib/python2.7/dist-packages/docker_registry/storage/__init__.py", line 36, in load
    _storage[kind] = engine.fetch(kind)(
  File "/usr/local/lib/python2.7/dist-packages/docker_registry/core/driver.py", line 267, in fetch
    % (name, name, available(), e)
NotImplementedError: You requested storage driver docker_registry.drivers.swift
which is not installed. Try `pip install docker-registry-driver-swift`
or check your configuration. The following are currently
available on your system: ['dumb', 'file', 's3']. Exception was: No module named swift
[2015-01-15 20:39:34 +0000] [13] [INFO] Worker exiting (pid: 13)
[2015-01-15 20:39:34 +0000] [20] [INFO] Booting worker with pid: 20
Traceback (most recent call last):
  File "/usr/local/bin/gunicorn", line 11, in <module>
    sys.exit(run())
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/wsgiapp.py", line 74, in run
    WSGIApplication("%(prog)s [OPTIONS] [APP_MODULE]").run()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/base.py", line 185, in run
    super(Application, self).run()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/app/base.py", line 71, in run
    Arbiter(self).run()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/arbiter.py", line 169, in run
    self.manage_workers()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/arbiter.py", line 477, in manage_workers
    self.spawn_workers()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/arbiter.py", line 542, in spawn_workers
    time.sleep(0.1 * random.random())
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/arbiter.py", line 209, in handle_chld
    self.reap_workers()
  File "/usr/local/lib/python2.7/dist-packages/gunicorn/arbiter.py", line 459, in reap_workers
    raise HaltServer(reason, self.WORKER_BOOT_ERROR)
gunicorn.errors.HaltServer: <HaltServer 'Worker failed to boot.' 3>
```

I can't understand why the driver seems to be "missing" when it's present in the system.

Can you help me?

Thanks in advance!
Andrés

**Comment by dmp42** [comment]
As far as I understand, you ran `pip install` on your host system, but are running the registry inside a docker container, correct?

**Comment by agonzalezandino** [comment]
Yeah, sorry about the misunderstanding. I ran the `pip install` in the host machine. The registry ran using the official container from: https://registry.hub.docker.com/u/library/registry/.

I made no changes to the container. The way I run it is explained in my original message.

**Comment by bacongobbler** [comment]
You need to install the library inside the container. For convenience: https://github.com/bacongobbler/docker-registry-driver-swift/blob/master/Dockerfile

https://registry.hub.docker.com/u/bacongobbler/registry-swift/ is probably what you'll want.

**Comment by bacongobbler** [comment]
To clarify, "`pip install` on the host machine" only applies when you are intending to run the registry as a process rather than a container, either for testing purposes or for other reasons. If you're running the container itself, it does not share the host filesystem so you'll have to install it inside there.

Does that help clarify things?

**Comment by agonzalezandino** [comment]
I see what you mean, but I'm not sure on how to implement it.

I can't completely understand the relation between the docker image in https://registry.hub.docker.com/u/bacongobbler/registry-swift/ and the repository https://github.com/bacongobbler/docker-registry-driver-swift/ that you mentioned.

Should I expect to get that exact image from building using the repository Dockerfile & assets? If that's the case, I don't see where `pip install docker-registry-driver-swift` fits.

Thank you!

**Comment by bacongobbler** [comment]
The image on dockerhub is an [automated build](http://docs.docker.com/docker-hub/builds/). Once a PR is merged into master, `bacongobbler/registry-swift` rebuilds itself from source. You should expect to get the same image if you built from source.

Considering that this is not an issue with the driver itself, I'm going to go ahead and close this.
